File: 0303-Range Sum Query - Immutable/c++/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class NumArray {
public:
    NumArray(vector<int>& nums) {
        dp = nums;
        for(int idx = 1; idx < nums.size(); ++idx) {
            dp[idx] += dp[idx-1];
        }
    }
    
    int sumRange(int i, int j) {
        if(i == 0) return dp[j];
        return dp[j] - dp[i-1];
    }
private:
    vector<int> dp;
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(i,j);
 */

// Runtime: 32 ms, faster than 99.87% of C++ online submissions for Range Sum Query - Immutable.
// Memory Usage: 17.3 MB, less than 48.05% of C++ online submissions for Range Sum Query - Immutable.

File: 0303-Range Sum Query - Immutable/golang/solution.go
package golang

type NumArray struct {
	DP []int
}

func Constructor(nums []int) NumArray {
	var numArray NumArray
	if len(nums) == 0 {
		return numArray
	}

	var dp []int
	dp = append(dp, nums[0])
	for idx := 1; idx < len(nums); idx++ {
		dp = append(dp, dp[idx-1]+nums[idx])
	}
	numArray.DP = dp
	return numArray
}

func (this *NumArray) SumRange(i int, j int) int {
	if i == 0 {
		return this.DP[j]
	}
	return this.DP[j] - this.DP[i-1]
}

/**
 * Your NumArray object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.SumRange(i,j);
 */

// Runtime: 32 ms, faster than 75.44% of Go online submissions for Range Sum Query - Immutable.
// Memory Usage: 12.1 MB, less than 5.26% of Go online submissions for Range Sum Query - Immutable.
